File: tests/machine/x/clj/group_by_having.clj
(ns main)

(defrecord _Group [key Items])

(defn _group_by [src keyfn]
  (let [groups (atom {})
        order (atom [])]
    (doseq [it src]
      (let [k (keyfn it)
            ks (str k)]
        (when-not (contains? @groups ks)
          (swap! groups assoc ks (_Group. k []))
          (swap! order conj ks))
        (swap! groups update ks (fn [g] (assoc g :Items (conj (:Items g) it)))))
    )
    (map (fn [k] (@groups k)) @order)))

(defn _escape_json [s]
  (-> s
      (clojure.string/replace "\\" "\\\\")
      (clojure.string/replace "\"" "\\\"")))

(defn _to_json [v]
  (cond
    (nil? v) "null"
    (string? v) (str "\"" (_escape_json v) "\"")
    (number? v) (str v)
    (boolean? v) (str v)
    (sequential? v) (str "[" (clojure.string/join "," (map _to_json v)) "]")
    (map? v) (str "{" (clojure.string/join "," (map (fn [[k val]]
                                        (str "\"" (_escape_json (name k)) "\":" (_to_json val))) v)) "}")
    :else (str "\"" (_escape_json (str v)) "\"")))

(defn _json [v]
  (println (_to_json v)))

(declare people big)

(defn -main []
  (def people [{:name "Alice" :city "Paris"} {:name "Bob" :city "Hanoi"} {:name "Charlie" :city "Paris"} {:name "Diana" :city "Hanoi"} {:name "Eve" :city "Paris"} {:name "Frank" :city "Hanoi"} {:name "George" :city "Paris"}]) ;; list of map of string to string
  (def big (map (fn [g] {:city (:key g) :num (count (:Items g))}) (_group_by people (fn [p] (:city p))))) ;; list of map of string to any
  (_json big)
)

(-main)

File: tests/transpiler/x/fs/group_by_having.fs
// Generated 2025-07-21 18:37 +0700

type Anon1 = {
    mutable name: string
    mutable city: string
}
type Anon2 = {
    mutable name: string
    mutable city: string
}
type Anon3 = {
    mutable city: obj
    mutable num: int
}
type Anon4 = {
    mutable p: obj
}
type Anon5 = {
    mutable key: obj
    mutable items: Anon4 list
}
type Anon6 = {
    mutable city: obj
    mutable num: int
}
let people: Anon2 list = [{ name = "Alice"; city = "Paris" }; { name = "Bob"; city = "Hanoi" }; { name = "Charlie"; city = "Paris" }; { name = "Diana"; city = "Hanoi" }; { name = "Eve"; city = "Paris" }; { name = "Frank"; city = "Hanoi" }; { name = "George"; city = "Paris" }]
let big: Anon6 list = [ for (key, items) in List.groupBy (fun p -> p.city) people do
    let g : Anon5 = { key = key; items = items }
    yield { city = g.key; num = List.length (g.items) } ]
json big
